src/librustdoc/clean/utils.rs

```diff
@@ -577,3 +579,37 @@ pub(super) fn render_macro_arms<'a>(
 pub(super) fn render_macro_matcher(matcher: &TokenTree) -> String {
     rustc_ast_pretty::pprust::tt_to_string(matcher)
 }
+
+pub(super) fn display_macro_source(
+    cx: &mut DocContext<'_>,
+    name: Symbol,
+    def: &ast::MacroDef,
+    def_id: DefId,
+    vis: impl Clean<Visibility>,
```

Comment:
You could avoid making this generic by passing in `Visibility` instead. The only drawback is it cleans the visibility early, instead of on-demand, but shrinking the generated code size seems worth it to me.

Reply:
I'd rather not, the conversion from the `ast::Visibility` needs to call `enter_resolver`, so if we can avoid it as much as possible, I think it's better. I can add a FIXME though if you want?

Reply:
Nah, that logic makes sense to me.